Marketplace business services: Feature branch

KokkieH opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
WooCommerce.com will be introducing a new type of product, Business Services, to the marketplace. (Internal post: p9h1Lv-c9Z-p2)

This PR adds support for including this new product type in the in-app marketplace at WooCommerce->Extensions.

Closes https://github.com/Automattic/woocommerce.com/issues/20444

This PR contains the changes from:

|PR|Description|
|---|---|
|#46530|Adds a Business Services tab to the marketplace.|
|#46725|Displays the new product type and categories on the new tab.|
|#47048|Fetch and display featured Business Service products on the Discover tab.|
|#47135|Add Business Services to the Search tab.|
|#47183|Add a new card type for the Business Services product type.|
|#47227|Ensure the Business Services tab only appears if there are products of that type on WooCommerce.com|
|#47352|Remove heading on Business Services tab, and some other small tweaks|
|#47452|Hide notices on Business Services and My Subscriptions tabs, and refactor search results to simplify component|

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prep

To test the request from your local WooCommerce core wp-env environment to your local WCCOM environment, you first need a running instance of WCCOM, as no products of the new type currently exists on production. 

1. Make sure your WCCOM environment is running on the latest version of the `feature/business-services-tab` branch.
2. Make sure you have a few products on the Business Services category, as well as its subcategories (see testing instructions for https://github.com/Automattic/woocommerce.com/pull/20275). 
3. Go to https://woocommerce.test/wp-admin/edit.php?post_type=featured_product and ensure that you have multiple featured products in different groups. Some of them should be in the Business Services group.
4. Check that the group is being returned from the featured endpoint https://woocommerce.test/wp-json/wccom-extensions/3.0/featured?country=US. (You'll probably need to flush the transient that caches this on your WCCOM with `wp cache flush`.)

Next you'll need to point your WC Core environment to your local WCCOM, as the `/wccom-extensions/3.0` API is not available yet on production:

1. Add a mu-plugin to your WC Core wp-env (as suggested in p6JqRr-aWb-p2), and put these two filters in it:

```php
add_filter( 'http_request_host_is_external', '__return_true' );

add_filter( 'http_request_args', function ( $r ) {
	$r['reject_unsafe_urls'] = false;
	$r['sslverify']          = false;
	$r['timeout']            = 60;

	return $r;
} );
```

2. Get the ID of wp-env's WordPress container.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/13509f82-962c-4f88-92ce-5f39222f2f64">

3. Link the container to woocommerce.test in the Docker network.

```bash
docker network connect woocommerce.test <container ID>
```

#### Test

1. Check out this branch
2. Edit `plugins/woocommerce/includes/admin/class-wc-admin-addons.php` and replace the start of the `fetch_featured()` function with the following:

```php
	public static function fetch_featured() {
		$transient_name = 'wc_addons_featured';
		// Important: WCCOM Extensions API v3.0 is used.
		$url      = 'https://woocommerce.test/wp-json/wccom-extensions/3.0/featured';
		$locale   = get_user_locale();
		$featured = false;
```

3. In `plugins/woocommerce-admin/client/marketplace/components/constants.ts`, edit the `MARKETPLACE_HOST` const to be `https://woocommerce.test`
4. Go to `/plugins/woocommerce` and run `pnpm --filter='@woocommerce/plugin-woocommerce' build`
5. Visit the [in-app marketplace](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions). Confirm a Business Services tab appears after a short delay. 
6. Confirm that a new section appears on the Discover tab containing the Business Service product you set as featured on WCCOM.
7. Switch to the Business Services tab. Confirm it shows your products, a list of Business Services child categories, and that you can filter the products by category.
8. Check the product cards against the designs: 5jyiYtoTF2ZUxgkulPVpOh-fi-3%3A3088
9. Open the dev console and run `localStorage.setItem( 'debug', 'wc-admin:*' );` Click a Business Service product card, then check that data for a Tracks event has been added to local storage (the dev env doesn't actually fire a Tracks event)
10. Use Search with a term that should return a Business Services product. If only Business Services are returned, you should see all matching products. If multiple product types are returned, they should be divided into plugin, theme and bs sections based on the product type.
11. Confirm that the other tabs also still look and work as expected.
12. Reverse the change made in step 3, and rebuild assets. Reload the marketplace. Now the Business Services tab should not appear (as you're querying production which doesn't have any of these products yet.

For more detailed instructions, please see the individual PRs.

<!-- End testing instructions -->
![Extensions_‹_WooCommerce_‹_woocommerce](https://github.com/woocommerce/woocommerce/assets/11873759/872e18b4-3317-471a-854c-9aa4c0da45e3)
![Extensions_‹_WooCommerce_‹_woocommerce](https://github.com/woocommerce/woocommerce/assets/11873759/f2cd41ca-9c0f-41df-9331-c57d4dee04e0)
![Extensions_‹_WooCommerce_‹_woocommerce](https://github.com/woocommerce/woocommerce/assets/11873759/723b3657-c0b1-41a0-b6b4-3deeefae7803)
![Extensions_‹_WooCommerce_‹_woocommerce](https://github.com/woocommerce/woocommerce/assets/11873759/b4b1e311-4abe-4323-8a7f-8cc53f356423)

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Add Business Services to the Extensions catalogue

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

KokkieH committed:
Add Business Services tab to in-app marketplace (#46530)

KokkieH committed:
Display Business Services products and categories in in-app marketplace (#46725)

Add support for a Business Services parent category to fetcCategories() function for in-app marketplace

raja651 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into feature/marketplace-business-services

raja651 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into feature/marketplace-business-services

KokkieH committed:
Add featured business services to Discover and other tabs (#47048)

* Update in-app marketplace to use wccom-extensions v3.0 API for featured endpoint

This lets us display featured Business Services in Discover tab and no-results state of Business Services tab

* Show featured business services in no-results view on Business Services tab

KokkieH committed:
Add business services products to in-app marketplace search results (#47135)

* Add Business Services section to search results in in-app marketplace

* Add fade-out to right of tab menu on mobile

- Gives some visual indication that there is a horizonal scroll

raja651 committed:
Add Business services card (#47183)

* Get new fields from extensions api and pass it to card component

* add business services card

* fix border for category badge and move businsess service card code to a func

* fix height issue for taller images for business service
- and add image transformation size for bigger images

* add product url and tracks to business services card

* removed unnecessary useeffect and add image transformation according to card height

* fix font sizes according to figma not keep it according to wccom business serivices card

KokkieH committed:
In app: hide business services if no products on WCCOM (#47227)

* Render Business Services tab conditionally

- Only show this tab if there are products of this type published on WCCOM

* change business service type in css for business services card

---------

Co-authored-by: Raja Sekar Manimaran <raja.sekar.manimaran@automattic.com>

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

KokkieH committed:
Hide Business Services tab heading in marketplace (#47352)

* Hide heading on Business Services tab

* Set _all category for tracks event for default Business Services tab view

* Business services - do not mention in search results if not needed

---------

Co-authored-by: Michal Iwanow <4765119+mcliwanow@users.noreply.github.com>

mcliwanow reviewed (approved):
I did another round of code review (not very thorough this time, as it was already reviewed in individual PRs) and tested the functionality with site connected to my local WooCommerce.test site.

Everything works as expected, considering that we'll merge https://github.com/woocommerce/woocommerce/pull/47452 too. 

It's ready to be shipped from my point of view.

KokkieH committed:
Hide notices on Business Services tab in marketplace (#47452)

* Do not display Woo Update Manager notice on Business Services or My Subscriptions tab - in-app marketplace

* Render <PluginInstallNotice/> inside <Content/> component instead of <Products/>

- Ensures notice is not rendered multiple times on Search tab

- Hide <ConnectNotice/> on Business Service and My Subscriptions tabs as well

* Refactor Search Results code to simplify and improve readability

- Reduce number of conditions we check and just render search result components selectively in final return instead

- Update search placeholder when there are no business services

---------

Co-authored-by: Michal Iwanow <4765119+mcliwanow@users.noreply.github.com>

Dan-Q commented on plugins/woocommerce-admin/client/marketplace/README.md:
```diff
@@ -9,13 +9,14 @@ The page contains two parts, the WooCommerce.com marketplace and a list of produ
 - **Discover**: A curated list of extensions and themes.
 - **Browse**: All extensions.
 - **Themes**: All themes.
+- **Business Services**: All business services
```
All the other items in this list end with a period.

```suggestion
- **Business Services**: All business services.
```

Dan-Q reviewed (approved):
Behaves as advertised; nice work. One non-blocking (one character) documentation suggestion.

KokkieH committed:
Update plugins/woocommerce-admin/client/marketplace/README.md

Fix documentation punctuation

Co-authored-by: Dan Q <dan@danq.me>

andfinally commented:
Just checking we're sizing the images as we intend – for themes, we do this on the container

```scss
overflow: hidden;
padding-top: 75%;
position: relative;
```

and this on the image

```scss
width: 100%;
position: absolute;
top: 0;
left: 0;
bottom: 0;
right: 0;
```

This constrains the image to 0.75 ratio while filling the container.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/8fcdf543-2b60-4cd4-aba0-820e9d51a77c">

I'm sure there's a more modern way to do this, e.g. `object-fit: cover;` but I'm wondering if it's intentional that the image shows up as contained in my browser.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/bbf8648a-2d75-4aed-8905-7e29353222c8">

andfinally commented:
Our conditional loading of the tab means there's a slight delay before it renders. This is better than showing an empty tab, and it's quite a slight effect, but maybe after this release we could look at making the tab permanent.

https://github.com/woocommerce/woocommerce/assets/1647564/b9d91be2-8ee8-4120-b98b-6baf00725043

KokkieH commented:
> Our conditional loading of the tab means there's a slight delay before it renders. It's not a massive problem, and it's better than showing an empty tab, but it would be worth making the tab permanent some time after this is released.

Yeah, this is unfortunately unavoidable atm--we can only decide whether or not to show the tab once we received search results from WCCOM, so either we only show it after a delay, or we show it and hide it after a delay.

Once we actually have products of this type on WCCOM we can definitely remove the selective rendering of this tab.

KokkieH committed:
Merge branch 'trunk' into feature/marketplace-business-services

KokkieH committed:
Merge branch 'trunk' into feature/marketplace-business-services
